### ISSUE TITLE
sorted realm results slow changeset processing down

### ISSUE THREAD

**Issue by klein-artur** [realm/realm-java#6614]
When I have a list with a bigger amount of elements (around 7000) and I add sorting to the query, change listeners are getting very slow. This is not the case when only one field is given. From the second field it slows down a lot. It also seems to depend on what field type is given to sorting. Booleans seem to be quite fast. Dates are also slower when only one is used.

Let me show you what I do: very simple query on all objects with multiple sorting elements:

```
val theList = realm.where(TheObject::class.java)
    .sort(
        arrayOf("fieldBool", "fieldDate", "fieldString"),
        arrayOf(Sort.DESCENDING, Sort.DESCENDING, Sort.DESCENDING)
    )
    .findAllAsync()

theList.addChangeListener { result: RealmResults<TheObject> ->
    // Set the list to the view.
}

```
Now, when some object in this resulting list is changed, the change listener needs a lot of time to update the data. Also other change listeners on the same table are very slow.

even if I sort the result after adding a changeListener (in the change lsitener):

```
val theList = realm.where(TheObject::class.java)
    .findAllAsync()

theList.addChangeListener { result: RealmResults<TheObject> ->
    viewToSet.list = result.sort(
        arrayOf("fieldBool", "fieldDate", "fieldString"),
        arrayOf(Sort.DESCENDING, Sort.DESCENDING, Sort.DESCENDING)
    )
}

```
the update happens very slow. I found out that it might has to do with calculating the changeset. If I add a changelistener to the realm instead of the result, it works quite fast. Still what is interesting is that I do not add a changeListener to the sorted list but to the unsorted list. sorting then works quite fast inside of the listener, but as soon as I have a sorted RealmResults living somewhere, the changeset creation seems to take a while... But only for that table. All other tables are working fast.

**Comment by cmelchior** [comment]
Do all the sorted fields have indexes?

**Comment by klein-artur** [comment]
@cmelchior Nope. Let me try this.

**Comment by klein-artur** [comment]
@cmelchior Having them changed to indexes does not help. I have the feeling it's a bit faster, still it takes a lot of time...

**Comment by klein-artur** [comment]
@cmelchior So what I can say is:
My feeling is that it depends on the fields I give for sorting. Boolean seem to work better than any other field.
Also this is not every time... I am clueless why...

**Comment by bmunkholm** [comment]
@klein-artur Could you be a bit more concrete what "slow" means?
Having a minimal repro case could also be helpful and make things a bit more concrete.

**Comment by klein-artur** [comment]
@cmelchior  If I say slow, I mean around 20 seconds.

I think I now can articulate what the problem is.

If a SORTED `RealmResults<TestData>` instance lives somewhere, independently where, all change listeners on any other `RealmResults<TestData>` or on single `TestData` need more time to calculate the changeset

Let me explain:

I created a testproject with a repo: 
https://github.com/klein-artur/realmChangeTest

This project has a very simple data structure so the time delays are not really big, but still good to describe. In this project, on first start, I add 1,000,000 elements to a list to have a lot of data. In my real project the data structure is quite complex, so the delays are a lot bigger.
Then there are three scenarios implemented (currently commented so you can try all three one by one). 
First one is without any sorting: Loading and listening on the million elements works very fast:
```
21:55:25.596 I/System.out: RealmTest: Fetch starts.
21:55:26.232 I/System.out: RealmTest: Data emitted.
21:55:39.450 I/System.out: RealmTest: Data clicked, Update commited.
21:55:39.751 I/System.out: RealmTest: Data emitted.
```
You see a delay of milliseconds. Everything nice.

Scenario 2 introduces sorting: Loading takes longer now (makes sense when sorting million elements) In my real project I have only 7000 elements so sorting works quite fast. Now a listener is added to this sorted list. The listener now also needs some time after the update to fire. So I assume calculating the changeset takes time on a sorted list. In my real project this changelistener now needs around 20 seconds to fire.
```
21:32:14.328 I/System.out: RealmTest: Fetch starts.
21:32:23.855 I/System.out: RealmTest: Data emitted.
21:33:03.000 I/System.out: RealmTest: Data clicked, Update commited.
21:33:10.894 I/System.out: RealmTest: Data emitted.
```

Now comes the interesting part (Scenario 3).
The list is fetched unsorded (works very fast again) and sorting is done before the list is set to the adapter (that sorting than takes a bit). 
The listener is added to the unsorted list, so one would assume that the listener might work fast now, because in the changeset of the change has nothing to do with sorting but:
```
22:04:45.312 I/System.out: RealmTest: Fetch starts.
22:04:45.906 I/System.out: RealmTest: Data emitted.
22:04:51.793 I/System.out: RealmTest: Data clicked, Update commited.
22:04:55.139 I/System.out: RealmTest: Data emitted.
```

Out of some reason, the changelistener added to this unsorted result now needs 3 seconds (this three seconds in my real project blow up to ca. 20 seconds). And not the sorting takes time, the print happens as the first line of the changelistener. The sort that takes around 7 seconds add on top of this.

So long story short:
If a SORTED `RealmResults<TestData>` instance lives somewhere, independently where, all change listeners on any other `RealmResults<TestData>` or on single `TestData` need more time to calculate the changeset. Complexitiy in data increases the time a lot.

When I have a look into the android profiler, I see that something is calculated on some background thread, but I cannot see what happens, so I'm quite sure that the changeset is calculated in native code.

**Comment by cmelchior** [comment]
Thanks, I'll take a look at your example.

How does your original data model look like?

Especially around relationships to other objects? And does it have cycles? We have some known edge cases in our current change detection algorithm that can cause it to become very slow in some cases. The solution here is usually to break up the forward relationships and replace them with `@LinkingObjects`, but if and how that is possible depends very much on the specific data model.

**Comment by klein-artur** [comment]
Unfortunately I cannot share the datastructure...

Yes, it has cycles: `ObjectA -> ObjectB -> List<ObjectA>` but they are never set. I think this was the solution before there where `@LinkingObjects`. I could test to break this but it will take a lot of efford. Maybe we can first find something different.

Still it seems that also uncycling databases do have that behavior (The example database has no relationships at all).

So I think the main question is:
Why does a sorted `RealmResults<T>` slow down all other listeners on `RealmResults<T>` or `T`, even if it has no own listener.

A solution for this could be to not calculate the changeset for results that have no changelistener attached. Because the calculation of changeset for sorted results seems to be the thing that exceeds in time (in the testproject the calculation of the changeset as long as no sorting happened takes a few milliseconds, but as soon as some sorted list is present takes around 3 seconds).

If the calculation of changeset happens only for Results that are listened, I could add a listener to the unsorted list and sort the list after I got the change.
